utils/weights.py

```diff
@@ -1,17 +1,174 @@
-        "preferences": 1,
-        "lowest fairness score": 1,
         "demand_deviation": 5,
+        "shift_duration": 15,
+
+        # Scaling excess and deficit of demand
+        "excess demand deviation factor": 1.0,
+        "deficit demand deviation factor": 1.0,
+
     }
 
 
-def get_shift_design_weights():
-    return {"demand_deviation": 5, "shift_duration": 15}
+def get_weights(time_set, staff):
+    """
+    The reference weight of demand deviation is added. Then the weights are scaled to hours, which is the reference
+    time step. Lastly, the weights are scaled relatively to each other and to the problem specific data.
+    """
+
+    weights = set_weights()
+    weights = correct_weights(weights)
+    weights = add_demand_deviation_weight(weights)
+    weights = scale_weights_to_hours(weights, time_set)
+    weights = scale_weights_relatively(weights, staff)
+    weights = scale_up_weights(weights, staff)
+
+    return weights
+
+
+def correct_weights(weights):
+    """
+    Weekly rest contributes positively to the fairness score (if additional hours are granted), different from the
+    other fairness aspects. Thus it is necessary to correct the weight, so that it actually represents the intention
+    defined in set_weights.
+
+    NOTE:   Preferences are also contributing positively to the fairness score, but these respective weights are
+            corrected naturally in the generate_preferences-function.
+    """
+
+    weights["rest"] = 1 / weights["rest"]
+
+    return weights
+
+
+def add_demand_deviation_weight(weights):
+    """
+    Weight of demand deviation is not an input value and is thus treated in a separate function. The weight is set
+    to 1, serving as the reference value for remaining weights. This value should not be changed.
+    """
+
+    weights["demand deviation"] = 1
+
+    return weights
+
+
+def scale_weights_to_hours(weights, time_set):
+    """
+    Demand deviation, contracted hours and preferences needs to be scaled to hours, as this is the reference time step.
+    """
+
+    time_step = time_set["step"]
+
+    weights["demand deviation"] *= time_step
+    weights["contracted hours"] *= time_step
+    weights["preferences"] *= time_step
+
+    return weights
+
+
+def scale_weights_relatively(weights, staff):
+
+    num_employees = len(staff["employees"])
+    employee_contracted_hours = staff["employee_contracted_hours"]
+
+    weights = scale_contracted_hours_relatively(weights, employee_contracted_hours)
+    weights = scale_least_favoured_employee(weights, num_employees)
+    weights = scale_demand_deviation(weights, num_employees)
+
+    return weights
+
+
+def scale_contracted_hours_relatively(weights, employee_contracted_hours):
+    """
+    Contracted hours weight is scaled for each employee based on the ratio between contracted hours and
+    default contracted hours. Scaling factor allows for
+    """
+
+    contracted_hours_weights = tupledict()
+    scaling_factor = 1.0
```

Comment:
Why is this needed? Doubling this is equivalent to just doubling the base contracted hours weight if I am not mistaken

Reply:
Experimented with different scaling of the weights, looked to change how the optimality model behaved. Included it, as it might be worth testing on a later stage.